Sources/GRPCProtobuf/InputParser.swift

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2024, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+import SwiftProtobuf
+import SwiftProtobufPluginLibrary
+
+import struct GRPCCodeGen.CodeGenerationRequest
+
+/// Parses a ``FileDescriptor`` object into a ``CodeGenerationRequest`` object.
+public struct InputParser {
+  public init() {}
+  public func parse(input: FileDescriptor) throws -> CodeGenerationRequest {
+    let fileName = input.name
+    let package = input.package
+    var documentation = String()
+    // Field number used to collect .proto file comments.
+    let documentationPath = IndexPath(index: 12)
+    if let syntaxLocation = input.sourceCodeInfoLocation(path: documentationPath) {
+      documentation = syntaxLocation.asSourceComment(
+        commentPrefix: "///",
+        leadingDetachedPrefix: "//"
+      )
+      // If the was a leading or tailing comment it won't have a blank
+      // line, after it, so ensure there is one.
+      if !documentation.isEmpty && !documentation.hasSuffix("\n\n") {
+        documentation.append("\n")
+      }
```

Comment:
This part is taken from here https://github.com/apple/swift-protobuf/blob/ea9ac1c4831d14118372ce7e1dcebd142c74e7f3/Sources/protoc-gen-swift/FileGenerator.swift#L71. Apparently there are fixed path numbers for different things that appear in a proto file - 12 is for the header. Should I add to NOTICES that I've used this small piece of code from SwiftProtobuf or should I add some link in the comment above the line containing the index?